### PR TITLE
[MISC] search algorithm refactoring

### PR DESCRIPTION
**Pull request by smehringer**
Of course the policies would be moved into their own files
and documentation is missing 
but the search_test runs through :tada:

### CODE REVIEW

**Comment by marehr** [comment]
Can you rebase?

File: include/seqan3/search/search.hpp [hunk 1]
```diff
@@ -89,6 +93,40 @@ struct search_configurator
...
+     */
+    template <typename configuration_t, typename index_t>
+    static auto configure_algorithm(configuration_t const & cfg, index_t const & index)
+    {
+        if constexpr (bi_fm_index_specialisation<index_t>)
+        {
+            using algorithm_t = search_scheme_algorithm<configuration_t,
+                                                        index_t,
+                                                        policy_max_error,
+                                                        policy_result_builder>;
```
Comment (marehr) [inline include/seqan3/search/search.hpp:118]: This shows that you didn't really use policies as intended.

Both classes use the same policies.

I'm not sure how all searches are, but from what I could see they all have an underlying search, like trivial search and bidirectional search (which could be policies) and you would get a single search_algorithm, that would look like this:

```c++
// pseudo code:
using algorithm_t = index_based_search_algorithm<
    configuration_t,
    index_t,
    // provides search_impl (currently search_trivial, search_algo_bi)
    is_bi_fm_index ? policy_search_scheme :
                     policy_unidirectional,
    // provides operator()
    switch(search_mode):
        mode == all : policy_search_mode_all
        mode == best_hits : policy_search_mode_best_hits
        mode == all_best_hits : policy_search_mode_all_best_hits
        mode == strata_hits : policy_search_mode_strata_hits,
    policy_max_error_state,
    policy_make_results>;
```

File: include/seqan3/search/detail/policy_max_error.hpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+     * If seqan3::max_error_rate is set in the configuration, the error rates are converted to error counts
+     * based on the length of the query sequence.
+     */
+    template <typename configuration_t, std::ranges::forward_range query_t>
+    auto get_search_param(configuration_t const & cfg, [[maybe_unused]] query_t && query)
+    {
+        using search_traits_t = search_traits<configuration_t>;
+
+        // retrieve error numbers / rates
+        detail::search_param max_error{0, 0, 0, 0};
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/search/detail/policy_max_error.hpp:45]: This renaming of an already existing struct can be done in another PR.

File: include/seqan3/search/detail/policy_max_error.hpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            // NOTE: Casting doubles rounds towards zero (i.e. floor for positive numbers). Thus, given a rate of
+            // 10% and a read length of 101 the max number of errors is correctly casted from 10.1 errors to 10
+            auto & [total, subs, ins, del] = max_error;
+            std::tie(total, subs, ins, del) = std::apply([& query] (auto && ... args)
+                {
+                    return std::tuple{(args * std::ranges::size(query))...};
+                }, get<search_cfg::max_error_rate>(cfg).value);
+        }
+
+        return max_error;
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/search/detail/policy_max_error.hpp:61]: But the config might not contain the `search_cfg::max_error_rate` configuration element, so I do need the `constexpr if else`. Or did I miss something in your suggested change?

File: include/seqan3/search/detail/policy_result_builder.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    auto make_results(std::vector<index_cursor_t> const & internal_hits, configuration_t const &)
+    {
+        using search_traits_t = search_traits<configuration_t>;
+
+        // output cursors or text_positions
+        if constexpr (search_traits_t::search_return_index_cursor)
+        {
+            return internal_hits;
+        }
+        else
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/search/detail/policy_result_builder.hpp:43]: What kind of splitting do you propose? (another make_results overload or two detail functions)
Note that the nesting depends on the `search_traits` which cannot be reflected in a parameter function overload.

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 5]
```diff
@@ -28,6 +28,112 @@ namespace seqan3::detail
...
+
+    //!\brief The configuration object.
+    configuration_t config{};
+
+    //!\brief A pointer to the fm index which is used to perform the unidirectional search.
+    index_t const * index_ptr{nullptr};
+
+    // forward declaration
+    template <bool abort_on_hit, typename query_t, typename delegate_t>
+    inline void search_algo_bi(query_t & query,
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/search/detail/search_scheme_algorithm.hpp:50]: Complaints can be sent to Chris, I'll take suggestions.

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 6]
```diff
@@ -28,6 +28,112 @@ namespace seqan3::detail
...
+        auto on_hit_delegate = [&internal_hits] (auto const & it)
+        {
+            internal_hits.push_back(it);
+        };
+
+        // choose mode
+        if constexpr (traits_t::search_best_hits || traits_t::search_all_best_hits || traits_t::search_strata_hits)
+        {
+            auto max_total = max_error.total;
+            max_error.total = 0; // start search with less errors
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/search/detail/search_scheme_algorithm.hpp:101]: alright

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 7]
```diff
@@ -28,6 +28,112 @@ namespace seqan3::detail
...
+                if (!internal_hits.empty())
+                {
+                    internal_hits.clear(); // TODO:don't clear when using Optimum Search Schemes with lower error bounds
+                    uint8_t const s = get<search_cfg::mode>(config).value;
+                    max_error.total += s - 1;
+                    search_algo_bi<false>(query, max_error, on_hit_delegate);
+                }
+            }
+        }
+        else // detail::search_mode_all
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/search/detail/search_scheme_algorithm.hpp:125]: In order to make those parts a policy, we would need to transform all policies into CRTP classes because they all need to call the private function `search_algo_bi`/`search_trivial`. I tried to avoid this as I think it's quite complicated.
Can this be part of a later refactoring?

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 8]
```diff
@@ -503,55 +607,33 @@ inline void search_ss(index_t const & index, query_t & query, search_param const
  * Strong exception guarantee if iterating the query does not change its state and if invoking the delegate also has a
  * strong exception guarantee; basic exception guarantee otherwise.
  */
-template <bool abort_on_hit, typename index_t, typename query_t, typename delegate_t>
-inline void search_algo_bi(index_t const & index, query_t & query, search_param const error_left,
-                           delegate_t && delegate)
+template <typename configuration_t, typename index_t, typename ...policies_t>
+template <bool abort_on_hit, typename query_t, typename delegate_t>
+inline void search_scheme_algorithm<configuration_t, index_t, policies_t...>::search_algo_bi(
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/search/detail/search_scheme_algorithm.hpp:612]: I don't understand. This basically is the `search_scheme_algorithm`. I just left the function (name) as it is and made it a private member function in `search_scheme_algorithm` in order to keep the structural changes in this file minimal.

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 9]
```diff
@@ -37,17 +41,127 @@ enum class error_type : uint8_t
...
+
+        // construct internal delegate for collecting hits for later filtering (if necessary)
+        std::vector<typename index_t::cursor_type> internal_hits;
+        delegate = [&internal_hits] (auto const & it)
+        {
+            internal_hits.push_back(it);
+        };
+
+        // choose mode
+        if constexpr (traits_t::search_best_hits || traits_t::search_all_best_hits || traits_t::search_strata_hits)
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/search/detail/unidirectional_search_algorithm.hpp:119]: As said above. Making this into a policy would require to make all policies into CRTP base classes and I tried to avoid this. Can this be part of a later refactoring?

File: include/seqan3/search/search.hpp (hunk 1 above)
Comment (smehringer) [reply include/seqan3/search/search.hpp:118]: When I read through the alignment algorithm and its policies and asked Rene about the design decisions, I thought to understand that policies are **reusable or customisable parts** of an algorithm. From this definition, I don't think search_trivial/unidirectional and search_algo_bi/search_scheme_algorithm should be designed as policies. They are the algorithms itself, no?
Additionally, I think that in the future, we might have actual differences between the search modi of the search algorithms uni/bi. So we would need to refactor this back into algorithms.

These are just my thoughts on this. Either way, could this be part of a later restructuring?

**Comment by smehringer** [comment]
Sorry, I had to rebase because of merge conflicts.

File: include/seqan3/search/detail/policy_max_error.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/search/detail/policy_max_error.hpp:62]: Oh, I see; I overlooked `get<search_cfg::max_error_rate>(cfg)`?

File: include/seqan3/search/detail/policy_max_error.hpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        }
+        else if constexpr (search_traits_t::search_with_max_error_rate)
+        {
+            // NOTE: Casting doubles rounds towards zero (i.e. floor for positive numbers). Thus, given a rate of
+            // 10% and a read length of 101 the max number of errors is correctly casted from 10.1 errors to 10
+            auto & [total, subs, ins, del] = max_error;
+            std::tie(total, subs, ins, del) = std::apply([& query] (auto && ... args)
+                {
+                    return std::tuple{(args * std::ranges::size(query))...};
+                }, get<search_cfg::max_error_rate>(cfg).value);
```
Comment (marehr) [inline include/seqan3/search/detail/policy_max_error.hpp:59]:
```suggestion
            std::tie(total, subs, ins, del) = std::apply([& query] (auto && ... args)
            {
                return std::tuple{(args * std::ranges::size(query))...};
            }, get<search_cfg::max_error_rate>(cfg).value);
```

File: include/seqan3/search/detail/policy_result_builder.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/search/detail/policy_result_builder.hpp:43]: Hmm? I think this should work:

```c++
    template <typename index_cursor_t, typename configuration_t>
        requires search_traits<configuration_t>::search_return_index_cursor
    auto make_results(std::vector<index_cursor_t> const & internal_hits, configuration_t const &)
```

or does it not work?

File: include/seqan3/search/detail/search_scheme_algorithm.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/search/detail/search_scheme_algorithm.hpp:125]: I don't think that you need a CRTP base class, because your policy `search_algo_bi` and `search_trivial` could have a single function called `underlying_search`.

I agree with postponing doing this as policies, but what limits you to split this into multiple smaller functions and delegate them in `operator()`?

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/search/detail/unidirectional_search_algorithm.hpp:119]: I doubt the CRTP class, but lets defer the policy.

I still would want the spilt, because this function is packed.

File: include/seqan3/search/search.hpp (hunk 1 above)
Comment (marehr) [reply include/seqan3/search/search.hpp:118]: A different behaviour for a mode could be implemented by a different underlying search via a policy which will be used conditionally. I find it funny, because from what I have seen, the underlying search is the **reusable or customisable part** of the algorithm. This might be a "Wo ein Wille ist, ist auch ein Weg" issue :)

File: include/seqan3/search/detail/policy_result_builder.hpp [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    template <typename index_cursor_t, typename configuration_t>
+    auto make_results(std::vector<index_cursor_t> const & internal_hits, configuration_t const &)
+    {
+        using search_traits_t = search_traits<configuration_t>;
+
+        // output cursors or text_positions
+        if constexpr (search_traits_t::search_return_index_cursor)
+        {
+            return internal_hits;
```
Comment (marehr) [inline include/seqan3/search/detail/policy_result_builder.hpp:41]: I just saw this; `internal_hits` is `const &`.

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 12]
```diff
@@ -28,6 +28,112 @@ namespace seqan3::detail
...
+     *
+     * \f$O(|query|^e)\f$ where \f$e\f$ is the total number of maximum errors.
+     */
+    template <typename query_t>
+    auto operator()(query_t && query) noexcept
+    {
+        auto error_state = this->max_error_counts(config, query); // see policy_max_error
+
+        // construct internal delegate for collecting hits for later filtering (if necessary)
+        std::vector<typename index_t::cursor_type> internal_hits;
```
Comment (marehr) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:91]: Here it is defined

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 13]
```diff
@@ -28,6 +28,112 @@ namespace seqan3::detail
...
+            }
+        }
+        else // detail::search_mode_all
+        {
+            // If you want to find all hits, you cannot stop once you found any hit (<false>)
+            // since you have to find all paths in the search tree that satisfy the hit condition.
+            search_algo_bi<false>(query, error_state, on_hit_delegate);
+        }
+
+        return this->make_results(internal_hits, config); // see search_result_builder_policy
```
Comment (marehr) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:132]: Here it is returned; return type is `const &`

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 14]
```diff
@@ -28,6 +28,112 @@ namespace seqan3::detail
...
+    /*!\brief Searches a query sequence in a bidirectional index.
+     * \tparam query_t Must model std::ranges::input_range over the index's alphabet.
+     * \param[in] query Query sequence to be searched in the index.
+     *
+     * ### Complexity
+     *
+     * \f$O(|query|^e)\f$ where \f$e\f$ is the total number of maximum errors.
+     */
+    template <typename query_t>
+    auto operator()(query_t && query) noexcept
```
Comment (marehr) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:86]: return type is `const &`

Isn't this a dangling reference?

File: include/seqan3/search/detail/policy_result_builder.hpp [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+protected:
+
+    /*!\brief Returns a range over index cursors or text positions, depending on the output configuration.
+     * \tparam index_cursor_t The type of index cursor used in the search algorithm.
+     * \tparam configuration_t The search configuration type.
+     * \param[in] internal_hits internal_hits A range over internal cursor results.
+     * \returns a range over index cursors or text positions, depending on the output configuration.
+     */
+    template <typename index_cursor_t, typename configuration_t>
+    auto make_results(std::vector<index_cursor_t> const & internal_hits, configuration_t const &)
```
Comment (marehr) [inline include/seqan3/search/detail/policy_result_builder.hpp:34]: I think this should get `internal_hits`  without `const &`.

File: include/seqan3/search/detail/search_scheme_algorithm.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/search/detail/search_scheme_algorithm.hpp:50]: I have no domain knowledge, I can only tell that this name is bad.

File: include/seqan3/search/detail/policy_result_builder.hpp [hunk 16]
```diff
@@ -24,52 +24,78 @@ struct policy_result_builder
...
+     * \tparam configuration_t The search configuration type.
+     * \param[in] internal_hits internal_hits A range over internal cursor results.
+     * \returns a range over text positions.
+     *
+     * \details
+     *
+     * The text positions are sorted and made unique before returning them.
+     */
+    template <typename index_cursor_t, typename configuration_t>
+    auto make_results(std::vector<index_cursor_t> internal_hits, configuration_t const &)
```
Comment (marehr) [inline include/seqan3/search/detail/policy_result_builder.hpp:81]: Can we add a `requires` here? It currently falls-through implicitly (if traits is not `search_return_index_cursor` and `search_best_hits`);

Is there maybe an explicit traits configuration which describes this case?

Or is this always the default? If so can you update the comment that says something like "The default behaviour is to return a range over text positions by calling locate on each cursor. If you want to limit the returned range or do not spent the time to locate on a cursor, specify ..."

Comment (smehringer) [reply include/seqan3/search/detail/policy_result_builder.hpp:81]: I added a requires for the output configuration (return_text_positions) and added documentation that this function is used for all search modi except search_best_hits (all, all_best, strata). I also added the requirement of return_text_positions to the `make_results` overload for search best hits. I hope this makes it clearer that the results depend on the output configuration as well as the search modus.

File: include/seqan3/search/detail/policy_result_builder.hpp [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+namespace seqan3::detail
+{
+
+//!\brief Provides the function `make_results` if inherited by a search algorithm.
+//!\ingroup search
+struct policy_result_builder
+{
+protected:
+
```
Comment (rrahn) [inline include/seqan3/search/detail/policy_result_builder.hpp:26]:
```suggestion
```

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 18]
```diff
@@ -28,6 +28,126 @@ namespace seqan3::detail
  * \{
  */
 
+/*!\brief The algorithm that performs a unidirectional search on an FM index using trivial backtracking.
+ * \tparam configuration_t The search configuration type.
+ * \tparam index_t The type of index; must model seqan3::bi_fm_index_specialisation.
+ */
+template <typename configuration_t, bi_fm_index_specialisation index_t, typename ...policies_t>
+class search_scheme_algorithm : public policies_t...
```
Comment (rrahn) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:36]:
```suggestion
class search_scheme_algorithm : protected policies_t...
```

File: include/seqan3/search/detail/policy_max_error.hpp [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+struct policy_max_error
+{
+protected:
+    /*!\brief Returns a detail::search_param object filled by the information from the configuration.
+     * \tparam configuration_t The search configuration type.
+     * \tparam query_t Must model std::ranges::foward_range over the index's alphabet.
+     * \param[in] cfg The configuration object.
+     * \param[in] query The current query sequence.
+     *
+     * If seqan3::max_error is set in the configuration, its value already is of type detail::search_param and
```
Comment (rrahn) [inline include/seqan3/search/detail/policy_max_error.hpp:34]:
```suggestion
     * \details
     * 
     * If seqan3::max_error is set in the configuration, its value already is of type detail::search_param and
```

File: include/seqan3/search/detail/policy_max_error.hpp [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+        if constexpr (search_traits_t::search_with_max_error)
+        {
+            return get<search_cfg::max_error>(cfg).value;
+        }
+        else if constexpr (search_traits_t::search_with_max_error_rate)
+        {
+            // NOTE: Casting doubles rounds towards zero (i.e. floor for positive numbers). Thus, given a rate of
+            // 10% and a read length of 101 the max number of errors is correctly casted from 10.1 errors to 10
+            auto & [total, subs, ins, del] = max_error;
+            std::tie(total, subs, ins, del) = std::apply([& query] (auto && ... args)
```
Comment (rrahn) [inline include/seqan3/search/detail/policy_max_error.hpp:56]:
```suggestion
            std::tie(total, subs, ins, del) = std::apply([&query] (auto && ... args)
```

File: include/seqan3/search/detail/policy_result_builder.hpp [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+        using position_t = std::conditional_t<index_t::text_layout_mode == text_layout::collection,
+                                              std::pair<typename index_t::size_type, typename index_t::size_type>,
+                                              typename index_t::size_type>;
+        std::vector<position_t> positions;
+
+        for (auto const & cur : internal_hits)
+        {
+            for (auto const & text_pos : cur.locate())
+                positions.push_back(text_pos);
+        }
```
Comment (rrahn) [inline include/seqan3/search/detail/policy_result_builder.hpp:103]:
```suggestion
            std::ranges::move(cursor.locate(), std::ranges::back_inserter(positions);
```

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 22]
```diff
@@ -28,6 +28,126 @@ namespace seqan3::detail
...
+    /*!\brief Constructs from a configuration object and an index.
+     * \tparam configuration_t The search configuration type.
+     * \tparam index_t The type of index; must model seqan3::bi_fm_index_specialisation.
+     * \param[in] cfg The configuration object that guides the search algorithm.
+     * \param[in] index The index used in the algorithm.
+     */
+    search_scheme_algorithm(configuration_t const & cfg, index_t const & index) : policies_t{}...
+    {
+        config = cfg;
+        index_ptr = &index;
```
Comment (rrahn) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:62]:
```suggestion
        index_ptr = std::addressof(index);
```

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 23]
```diff
@@ -28,6 +28,126 @@ namespace seqan3::detail
...
+    }
+    //!\}
+
+    /*!\brief Searches a query sequence in a bidirectional index.
+     * \tparam query_t Must model std::ranges::input_range over the index's alphabet.
```
Comment (rrahn) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:67]:
```suggestion
     * \tparam query_t The type of the query sequence; must model std::ranges::input_range over the index's alphabet.
```

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 24]
```diff
@@ -28,6 +28,126 @@ namespace seqan3::detail
...
+     *
+     * \f$O(|query|^e)\f$ where \f$e\f$ is the total number of maximum errors.
+     */
+    template <typename query_t>
+    auto operator()(query_t && query) noexcept
+    {
+        auto error_state = this->max_error_counts(config, query); // see policy_max_error
+
+        // construct internal delegate for collecting hits for later filtering (if necessary)
+        std::vector<typename index_t::cursor_type> internal_hits;
```
Comment (rrahn) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:80]:
```suggestion
        std::vector<typename index_t::cursor_type> internal_hits{};
```

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 25]
```diff
@@ -28,6 +28,126 @@ namespace seqan3::detail
...
+            {
+                // * If you only want the best hit (traits_t::search_best_hits), you stop after finding the
+                //   first hit, the hit with the least errors (`abort_on_hit` is true).
+                // * If you are in strata mode (traits_t::search_strata_hits), you do the same as with best hits,
+                //   but then do the extra step afterwards (`abort_on_hit` is true).
+                // * If you want all best hits (traits_t::search_all_best_hits), you do not stop after the first
+                //   hit but continue the current search algorithm/max_error pattern (`abort_on_hit` is true).
+                constexpr bool abort_on_hit = !traits_t::search_all_best_hits;
+                search_algo_bi<abort_on_hit>(query, error_state, on_hit_delegate);
+                error_state.total++;
```
Comment (rrahn) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:129]:
```suggestion
                ++error_state.total;
```

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 26]
```diff
@@ -16,9 +16,13 @@
 #include <type_traits>
 
 #include <seqan3/alphabet/concept.hpp>
+#include <seqan3/core/detail/test_accessor.hpp>
+#include <seqan3/core/type_traits/deferred_crtp_base.hpp>
```
Comment (rrahn) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:20]: not needed?

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 27]
```diff
@@ -16,9 +16,13 @@
...
 #include <seqan3/range/concept.hpp>
 #include <seqan3/range/views/drop.hpp>
+#include <seqan3/search/detail/search_traits.hpp>
```
Comment (rrahn) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:23]: order

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 28]
```diff
@@ -37,17 +41,144 @@ enum class error_type : uint8_t
...
+     */
+    unidirectional_search_algorithm(configuration_t const & cfg, index_t const & index) : policies_t{}...
+    {
+        config = cfg;
+        index_ptr = &index;
+    }
+    //!\}
+
+    /*!\brief Searches a query sequence in an FM index using trivial backtracking.
+     * \tparam query_t       Must model std::ranges::input_range over the index's alphabet.
```
Comment (rrahn) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:80]:
```suggestion
     * \tparam query_t The type of the query sequence to search; must model std::ranges::input_range over the index's alphabet.
```

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 29]
```diff
@@ -37,17 +41,144 @@ enum class error_type : uint8_t
...
+     * \param[in] query      Query sequence to be searched in the index.
```
Comment (rrahn) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:81]:
```suggestion
     * \param[in] query Query sequence to be searched in the index.
```

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 30]
```diff
@@ -37,17 +41,144 @@ enum class error_type : uint8_t
...
+     *
+     * \f$O(|query|^e)\f$ where \f$e\f$ is the maximum number of errors.
+     */
+    template <typename query_t>
+    auto operator()(query_t && query) noexcept
+    {
+        auto error_state = this->max_error_counts(config, query); // see policy_max_error
+
+        // construct internal delegate for collecting hits for later filtering (if necessary)
+        std::vector<typename index_t::cursor_type> internal_hits;
```
Comment (rrahn) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:93]:
```suggestion
        std::vector<typename index_t::cursor_type> internal_hits{};
```

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 31]
```diff
@@ -37,17 +41,144 @@ enum class error_type : uint8_t
...
+            {
+                // * If you only want the best hit (traits_t::search_best_hits), you stop after finding the
+                //   first hit, the hit with the least errors (`abort_on_hit` is true).
+                // * If you are in strata mode (traits_t::search_strata_hits), you do the same as with best hits,
+                //   but then do the extra step afterwards (`abort_on_hit` is true).
+                // * If you want all best hits (traits_t::search_all_best_hits), you do not stop after the first
+                //   hit but continue the current search algorithm/max_error pattern (`abort_on_hit` is true).
+                constexpr bool abort_on_hit = !traits_t::search_all_best_hits;
+                search_trivial<abort_on_hit>(index_ptr->cursor(), query, 0, error_state, error_type::none);
+                error_state.total++;
```
Comment (rrahn) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:148]:
```suggestion
                ++error_state.total;
```

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 32]
```diff
@@ -37,17 +41,144 @@ enum class error_type : uint8_t
...
+            }
+
+            if constexpr (traits_t::search_strata_hits)
+            {
+                if (!internal_hits.empty())
+                {
+                    internal_hits.clear();
+                    uint8_t const s = get<search_cfg::mode>(config).value;
```
Comment (rrahn) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:156]: can we have a better name for `s`?

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 33]
```diff
@@ -37,17 +41,144 @@ enum class error_type : uint8_t
...
+
+private:
+    //!\brief The configuration object.
+    configuration_t config{};
+
+    //!\brief A pointer to the fm index which is used to perform the unidirectional search.
+    index_t const * index_ptr{nullptr};
+
+    //!\brief A function object that stores the on-hit-delegate to be executed whenever a hit in the index is found.
+    std::function<void(typename index_t::cursor_type const &)> delegate;
```
Comment (rrahn) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:112]: just out of curiosity, why is this needed? In the other implementation you pass it into the perform_search function, which seems plausible to do.

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 34]
```diff
@@ -28,6 +28,126 @@ namespace seqan3::detail
...
+            }
+            if constexpr (traits_t::search_strata_hits)
+            {
+                if (!internal_hits.empty())
+                {
+                    internal_hits.clear(); // TODO:don't clear when using Optimum Search Schemes with lower error bounds
+                    uint8_t const s = get<search_cfg::mode>(config).value;
+                    error_state.total += s - 1;
```
Comment (rrahn) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:137]: can we have a better name for `s`?

Comment (smehringer) [reply include/seqan3/search/detail/search_scheme_algorithm.hpp:137]:
```suggestion
                    error_state.total += stratum - 1;
```

File: include/seqan3/search/detail/search_scheme_algorithm.hpp [hunk 35]
```diff
@@ -28,6 +28,126 @@ namespace seqan3::detail
...
+                constexpr bool abort_on_hit = !traits_t::search_all_best_hits;
+                search_algo_bi<abort_on_hit>(query, error_state, on_hit_delegate);
+                error_state.total++;
+            }
+            if constexpr (traits_t::search_strata_hits)
+            {
+                if (!internal_hits.empty())
+                {
+                    internal_hits.clear(); // TODO:don't clear when using Optimum Search Schemes with lower error bounds
+                    uint8_t const s = get<search_cfg::mode>(config).value;
```
Comment (smehringer) [inline include/seqan3/search/detail/search_scheme_algorithm.hpp:136]:
```suggestion
                    uint8_t const stratum = get<search_cfg::mode>(config).value;
```

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp (hunk 26 above)
Comment (smehringer) [reply include/seqan3/search/detail/unidirectional_search_algorithm.hpp:20]:
```suggestion
```

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp (hunk 32 above)
Comment (smehringer) [reply include/seqan3/search/detail/unidirectional_search_algorithm.hpp:156]:
```suggestion
                    uint8_t const stratum = get<search_cfg::mode>(config).value;
```

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp [hunk 36]
```diff
@@ -37,17 +41,144 @@ enum class error_type : uint8_t
...
+                error_state.total++;
+            }
+
+            if constexpr (traits_t::search_strata_hits)
+            {
+                if (!internal_hits.empty())
+                {
+                    internal_hits.clear();
+                    uint8_t const s = get<search_cfg::mode>(config).value;
+                    error_state.total += s - 1;
```
Comment (smehringer) [inline include/seqan3/search/detail/unidirectional_search_algorithm.hpp:157]:
```suggestion
                    error_state.total += stratum - 1;
```

File: include/seqan3/search/detail/unidirectional_search_algorithm.hpp (hunk 33 above)
Comment (smehringer) [reply include/seqan3/search/detail/unidirectional_search_algorithm.hpp:111]: I was trying to reduce the parameter list of the free functions. search_trivial is fine but when you look into the free functions in `search_scheme_algorithm.hpp` (the ones I haven't touched yet) the number of parameters is huge. So this was rather a trial here, as I thought it is equivalent, but might serve as an example for further refactoring in  `search_scheme_algorithm.hpp`. If it is not equivalent but the function parameter is preferable I have no strong feelings to change it back.
